eis_toolkit/prediction/weights_of_evidence.py

```diff
@@ -0,0 +1,255 @@
+
+    Args:
+        evidential_raster: The evidential raster.
+        deposits: Vector data representing the mineral deposits or occurences point data.
+        raster_nodata: If nodata value of raster is wanted to specify manually. Optional parameter, defaults to None
+            (nodata from raster metadata is used).
+        weights_type: Accepted values are 'unique' for unique weights, 'ascending' for cumulative ascending weights,
+            'descending' for cumulative descending weights. Defaults to 'unique'.
+        studentized_contrast_threshold: Studentized contrast threshold value used to reclassify all classes.
+            Reclassification is used when creating generalized rasters with cumulative weight type selection.
+            Not needed if weights_type is 'unique'. Defaults to 2.
+        rasters_to_generate: Rasters to generate from the computed weight metrics. All column names
+            in the produced weights_df are valid choices. Defaults to ["Class", "W+", "S_W+]
+            for "unique" weights_type and ["Class", "W+", "S_W+", "Generalized W+", "Generalized S_W+"]
+            for the cumulative weight types.
+
+    Returns:
+        Dataframe with weights of spatial association between the input rasters.
+        Dictionary of output raster arrays.
+        Raster metadata.
+    """
+    out_df_col_names = [
+        "Class",
+        "Pixel count",
+        "Deposit count",
+        "W+",
+        "S_W+",
+        "W-",
+        "S_W-",
+        "Contrast",
+        "S_Contrast",
+        "Studentized contrast",
+    ]
+    if rasters_to_generate is not None and not all(col_name in rasters_to_generate for col_name in out_df_col_names):
+        raise exceptions.InvalidColumnException("Rasters to generate list contains metrics / column names.")
+
+    metrics_to_rasters = rasters_to_generate
+    if weights_type != "unique" and rasters_to_generate == ["Class", "W+", "S_W+"]:
+        metrics_to_rasters += ["Generalized W+", "Generalized S_W+"]
+
+    # 1. Data preprocessing
+    evidence_array = _read_and_preprocess_evidence(evidential_raster, raster_nodata)
+    raster_meta = evidential_raster.meta
+
+    # Rasterize deposits
+    deposit_array, _ = rasterize_vector(
+        geodataframe=deposits, default_value=1.0, base_raster_profile=raster_meta, fill_value=0.0
+    )
+
+    # Mask NaN out of the array
+    nodata_mask = np.isnan(evidence_array)
+    masked_evidence_array = evidence_array[~nodata_mask]
+    masked_deposit_array = deposit_array[~nodata_mask]
+
+    # 2. WofE calculations
+    if weights_type == "unique":
+        wofe_weights = _unique_weights(masked_deposit_array, masked_evidence_array)
+    elif weights_type == "ascending":
+        wofe_weights = _cumulative_weights(masked_deposit_array, masked_evidence_array, ascending=True)
+    elif weights_type == "descending":
+        wofe_weights = _cumulative_weights(masked_deposit_array, masked_evidence_array, ascending=False)
```

Comment:
```suggestion
    if weights_type == "unique":
        wofe_weights = _unique_weights(masked_deposit_array, masked_evidence_array)
    elif weights_type == "ascending":
        wofe_weights = _cumulative_weights(masked_deposit_array, masked_evidence_array, ascending=True)
    elif weights_type == "descending":
        wofe_weights = _cumulative_weights(masked_deposit_array, masked_evidence_array, ascending=False)
    else:
        raise ValueError("Expected weights_type to be one of unique, ascending or descending.")
```

Reply:
The execution should never go to the else-clause in this case since beartype checks `weights_type`.

Reply:
I almost take the cautious approach when dealing with Python, you never know what might go through some function :D.